packages/driver/src/cypress/log.ts

```diff
@@ -349,33 +349,62 @@ export class Log {
       next: null,
     })
 
-    const snapshot = this.cy.createSnapshot(name, this.get('$el'))
+    let snapshot
+    const postProcessSnapshot = (shouldRebindSnapshotFn = true) => {
+      const snapshots = this.get('snapshots') || []
+
+      // don't add snapshot if we couldn't create one, which can happen
+      // if the snapshotting process errors
+      // https://github.com/cypress-io/cypress/issues/15816
+      if (snapshot) {
+      // insert at index 'at' or whatever is the next position
+        snapshots[options.at || snapshots.length] = snapshot
+      }
 
-    const snapshots = this.get('snapshots') || []
+      this.set('snapshots', snapshots)
 
-    // don't add snapshot if we couldn't create one, which can happen
-    // if the snapshotting process errors
-    // https://github.com/cypress-io/cypress/issues/15816
-    if (snapshot) {
-      // insert at index 'at' or whatever is the next position
-      snapshots[options.at || snapshots.length] = snapshot
+      if (options.next && shouldRebindSnapshotFn) {
+        const originalLogSnapshotFn = this.snapshot
+
+        this.snapshot = function () {
+        // restore the original snapshot function
+          this.snapshot = originalLogSnapshotFn
+
+          // call orig fn with next as name
+          return originalLogSnapshotFn.call(this, options.next)
+        }
+      }
+
+      return this
     }
 
-    this.set('snapshots', snapshots)
+    if (this.config('experimentalSessionAndOrigin') && !Cypress.isCrossOriginSpecBridge) {
+      // @ts-ignore
+      const activeSpecBridgeOriginPolicyIfApplicable = this.state('currentActiveOriginPolicy') || undefined
+      // @ts-ignore
+      const { originPolicy: originPolicyThatIsSoonToBeOrIsActive } = Cypress.Location.create(this.state('anticipatingCrossOriginResponse')?.href || this.state('url'))
 
-    if (options.next) {
-      const fn = this.snapshot
+      if (activeSpecBridgeOriginPolicyIfApplicable && activeSpecBridgeOriginPolicyIfApplicable === originPolicyThatIsSoonToBeOrIsActive) {
+        // NOTE: using a combination of 'id' and date epoch to generate a unique event to prevent any type of doubling up for log requests for snapshots
+        const eventID = this.get('id') + Date.now()
 
-      this.snapshot = function () {
-        // restore the fn
-        this.snapshot = fn
+        Cypress.primaryOriginCommunicator.once(`snapshot:for:log:generated:${eventID}`, (generatedCrossOriginSnapshot) => {
+          snapshot = generatedCrossOriginSnapshot.body ? generatedCrossOriginSnapshot : null
+          postProcessSnapshot(false)
+        })
 
-        // call orig fn with next as name
-        return fn.call(this, options.next)
+        Cypress.primaryOriginCommunicator.toSpecBridge(activeSpecBridgeOriginPolicyIfApplicable, 'generate:snapshot:for:log', {
+          name,
+          id: eventID,
```

Comment:
Why can't the log's id be used? I'd assume we can associate multiple snapshots to a log so I would expect this to be fine / no need to manage another unique eventID.

Reply:
After revisiting, I removed the `date.now()` and opted for just going with the log ID, which the event handling now lives in the event manager to ask the spec bridge for the log snapshot.